File: chapter-02/linked-lists/linkedlists.go
package main

import (
	"errors"
	"fmt"
)

type Node struct {
	Data int
	Next *Node
}

type LinkedList struct {
	Head   *Node
	Tail   *Node
	Length int
}

func (ll *LinkedList) InsertAt(index, data int) error {
	if index < 0 || index >= ll.Length {
		return errors.New("index out of range")
	}

	newNode := Node{data, nil}

	if index == 0 {
		newNode.Next = ll.Head
		ll.Head = &newNode
	} else if index == ll.Length-1 {
		ll.Tail.Next = &newNode
		ll.Tail = &newNode
	} else {
		currentNode := ll.Head
		for range index - 1 {
			currentNode = currentNode.Next
		}

		newNode.Next = currentNode.Next
		currentNode.Next = &newNode
	}

	ll.Length++

	return nil
}

func (ll *LinkedList) AddToFront(data int) {
	newNode := Node{data, nil}
	if ll.Head == nil {
		ll.Head = &newNode
		ll.Tail = &newNode
	} else {
		newNode.Next = ll.Head
		ll.Head = &newNode
	}

	ll.Length++
}

func (ll *LinkedList) AddToBack(data int) {
	newNode := Node{data, nil}
	if ll.Head == nil {
		ll.Head = &newNode
		ll.Tail = &newNode
	} else {
		ll.Tail.Next = &newNode
		ll.Tail = &newNode
	}

	ll.Length++
}

func (ll *LinkedList) Remove(index int) error {
	if index < 0 || index >= ll.Length {
		return errors.New("index out of range")
	}

	if index == 0 {
		ll.Head = ll.Head.Next
		if ll.Head == nil {
			ll.Tail = nil
		}
	} else if index == ll.Length-1 {
		currentNode := ll.Head
		for range index - 1 {
			currentNode = currentNode.Next
		}
		currentNode.Next = nil
		ll.Tail = currentNode
	} else {
		currentNode := ll.Head
		for range index - 1 {
			currentNode = currentNode.Next
		}
		currentNode.Next = currentNode.Next.Next
	}

	ll.Length--
	return nil
}

func main() {
	linkedList := LinkedList{}
	for i := range 5 {
		if i == 0 {
			continue
		}
		linkedList.AddToBack(i)
	}

	currentNode := linkedList.Head
	for i := 0; i < linkedList.Length; i++ {
		fmt.Println(currentNode.Data)
		currentNode = currentNode.Next
	}

	fmt.Println()

	err := linkedList.Remove(1)
	if err != nil {
		panic(err)
	}

	currentNode = linkedList.Head
	for i := 0; i < linkedList.Length; i++ {
		fmt.Println(currentNode.Data)
		currentNode = currentNode.Next
	}

	err = linkedList.InsertAt(1, 90)
	if err != nil {
		panic(err)
	}

	fmt.Println()

	currentNode = linkedList.Head
	for i := 0; i < linkedList.Length; i++ {
		fmt.Println(currentNode.Data)
		currentNode = currentNode.Next
	}
}

File: chapter-02/linked-lists/linked_lists.py
class Node:
    def __init__(self, data, next) -> None:
        self.data: int = data
        self.next: Node = next
    

    def __str__(self):
        return str(self.data)


class LinkedList:
    def __init__(self) -> None:
        self.head = None
        self.tail = None


    def add_to_front(self, data: int) -> None:
        new_node = Node(data, None)
        if self.head == self.tail == None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node


    def add_to_back(self, data: int) -> None:
        new_node = Node(data, None)
        if self.head == self.tail == None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node


    def remove(self, index: int) -> None:
        if 0 > index or index >= self.length:
            raise IndexError()
        
        if index == 0:
            if self.head.next:
                self.head = self.head.next
            else:
                self.head = self.tail = None
        elif index == self.length - 1:
            current_node = self.head
            for _ in range(index - 1):
                current_node = current_node.next
            current_node.next = None
            self.tail = current_node
        else:
            current_node = self.head
            for i in range(index):
                if i == index - 1:
                    current_node.next = current_node.next.next
                    break
                current_node = current_node.next


    def insert_at(self, index: int, data: int) -> None:
        if 0 > index or index >= self.length:
            raise IndexError()
        
        new_node = Node(data, None)
        if index == 0:
            new_node.next = self.head
            self.head = new_node
        elif index == self.length - 1:
            self.tail.next = new_node
            self.tail = new_node
        else:
            for i, node in enumerate(self.__iter__()):
                if i == index - 1:
                    new_node.next = node.next
                    node.next = new_node
        

    @property
    def length(self) -> int:
        current_node = self.head
        if current_node == None:
            return 0
        
        counter = 1
        while current_node.next != None:
            counter += 1
            current_node = current_node.next

        return counter
    
    
    def __iter__(self) -> iter:
        current_node: Node = self.head
        while current_node:
            yield current_node
            current_node = current_node.next


    def __str__(self) -> str:
        return str([node.data for node in self.__iter__()])


if __name__ == "__main__":
    ll = LinkedList()

    ll.add_to_front(5)
    ll.add_to_front(4)
    ll.add_to_front(3)
    ll.add_to_front(2)
    ll.add_to_front(1)

    ll.remove(4 )

    print(ll.__str__())
    